service/storage_service.hh

```diff
@@ -760,6 +760,12 @@ private:
 public:
     future<bool> is_cleanup_allowed(sstring keyspace);
     bool is_repair_based_node_ops_enabled(streaming::stream_reason reason);
+
+private:
+    std::unordered_set<gms::inet_address> normal_state_handled;
```

Comment:
What if we replace with the same ip?

Reply:
I only ever see nodes added to this list, who cleans it up?

Reply:
> What if we replace with the same ip?

We do not care about the same ip, because the node won't wait for itself to become normal. The node only cares other nodes in the cluster that could recreate the connection.

Reply:
> I only ever see nodes added to this list, who cleans it up?

We could clear it up after the node has finished using it during the bootstrap process. I think this might be useful after bootstrap so I did not do that. It won't be a big problem even if we do not clear it and we have 1000 nodes in the cluster.

Reply:
> > What if we replace with the same ip?
> 
> We do not care about the same ip, because the node won't wait for itself to become normal. The node only cares other nodes in the cluster that could recreate the connection.

Node A waited for not B to become normal. Entered its IP into the map. B was replace by C with the same IP. A sees it as normal.

Reply:
In this case, the bootstrap ops will abort because B is gone. In addition, replace ops is not allowed because the bootstrap is in progress.

Reply:
I do not understand. B is being replaced by C which is also is kind of bootstrap. Or do you check the map only in the bootstrap path and not the replace one? Can replace have the same problem?

Reply:
Note, this check happens only during the boot up process. After the bootstrap is done. We do not care the value in the map. 

In your example, after A bootstrap ops, yes, C could replace B, but A does not care that anymore. 

In theory, the connection close issue could happen to replace as well. We do not see this in the tests. RBNO for replace is enabled by default for a long time.

Reply:
> 
> In theory, the connection close issue could happen to replace as well. We do not see this in the tests. RBNO for replace is enabled by default for a long time.

We should not wait for it to happen. We need to understand if it can happen or not and fix it if it can. May be because when node is replaced its topology is knows the issue cannot happen there?

Reply:
Nobody is waiting for it to happen. I am stating the fact we never saw it in practice. In v2 of this PR, we already have a patch to wait for replace ops as well.

Reply:
Cool.